Content: Logic module (#1341)
CodeYourFuture/curriculum: common-content/en/module/logic/search/index.md

+++
title = "1000 Bottles of Wine"
time = 60
emoji= "🐀"
[build]
render = 'never'
list = 'local'
publishResources = false
+++

> Now we're going to solve a puzzle together. We're going to think through a few different ways to approach it. This is a well-known puzzle and you may certainly google the answer. However, you will steal from yourself the opportunity of learning this important idea, so maybe _don't_ do that.

Reviewer: ```suggestion
Now we're going to solve a puzzle together. We're going to think through a few different ways to approach it. This is a well-known puzzle and it's certainly possible to google the answer. However, we'd suggest: don't do that - it would steal from yourself the opportunity of learning this important idea for yourself.
```

With 500 rats, it's similar, but not the same. Divide the 1000 bottles into 500 groups of 2. In round 1, give each rat a sip from one bottle in each group. Wait one hour. If the rat is alive, the bottle is safe. If the rat is dead, you found the poisoned bottle. If all rats are alive, the poisoned bottle is in the 500 bottles that were not tested. You still have 500 rats, so you can have each rat test the other bottle in their group.

This solution will take a maximum of 2 hours.

{{</tabs>}}

> #### New condition
>
> You now have just **10** rats to test the wine.

{{<tabs name="decimation">}}
===[[Write down your solution]]===

In your notebook, write down your solution to this problem. Here are the constraints again:

🍾 You have **1000** wine bottles, one of which is poisoned.  
🐀 You have **10** rats you can use to test the wine.  
⌛ The poison is lethal and will kill within **1** hour.

===[[Look at a solution]]===

> With just 10 rats, we need a way to make our problem simpler. Think of it as a search pattern, narrowing down our suspects with each round of testing.

Reviewer: I'm not sure I'd characterise the improvement as simplifying? We need a more _efficient_ solution maybe?

**Round 1:** Let's divide our 1000 bottles into 10 groups of 100. Each rat tests one group. An hour later, with one dead rat, we have narrowed our search to 100 bottles. We've saved 900 bottles of wine and we still have 9 rats to help us.

> **Round 2** Let's do it again: take our suspicious 100 bottles and create ten groups of ten. Our doomed rat perishes again, and after another hour, we've found the right group of ten. We now have 8 rats.

Reviewer: ```suggestion
**Round 2** Let's do it again: take our suspicious 100 bottles and create ten groups of ten. Our nine rats each drink from one whole group (and one group goes un-drunk). After another hour, either a rat dies and we're down to ten bottles to investigate, or no rat died and we know the un-drunk group contains the poison. We now have at least 8 rats.
```

Author: I will take these changes but rewrite because these sentences are too convoluted. But I take the point

**Round 3** Now we're down to just 10 bottles. Eight rats test a bottle each, leaving two untested. If a rat dies, we've found the poisoned bottle. If not, we have two bottles left for the last round.

**Round 4: Base case** We can use just one rat to test one bottle. If we have a dead rat, it's the poisoned bottle. If not, the last untasted bottle is the culprit.

Through careful division and patience, we've turned an overwhelming search into a manageable investigation. Four hours, and we'll have our answer.

{{</tabs>}}

> #### New condition
>

> > The birthday party is now in **1** hour, so you only have time for one test!

Reviewer: I think we should frame this slightly differently - we're really asking a different question here.

My initial question was: How many rats do we have in this iteration?

But we're not playing the same game - now we're playing a game of "how many rats would we need" rather than "how do we use the number of rats we have" - let's frame it that way, rather than saying "Same game: New condition" which I think we're currently doing?

Author: I think it's the same game as I'm still giving them 10 rats. They only need 7 in the end, which we can hammer down on afterwards? I only want to add one more constraint, not completely change the game, to provide a bridge from the last approach.

{{<tabs name="binary">}}
===[[Write down your solution]]===
Think about how you can use some of the thinking tools you just learned to solve this problem.

You will need boolean logic, deduction, and binary at least.

===[[Look at a solution]]===

If we only have one hour, we can only do a single round of testing! Our strategy of narrowing down can't work. We're going to have to think about what we have learned about binary to solve this problem.

I really advise that you write this out yourself in binary as you follow the solution through.

Let's reduce our case. Let's say we have 3 bottles and 2 rats.

We're going to number the bottles in binary. Each rat will test a different _bit_ of the binary number. Rat 1 will test the 1s place, Rat 2 the 2s place.

| Decimal | Binary | Rat 1 | Rat 2 |
| ------- | ------ | ----- | ----- |
| 1       | 001    | x     |       |
| 2       | 010    |       | x     |
| 3       | 011    | x     | x     |

> If rat 1 dies, we know the poison is in the 1s place. It's bottle 1. If rat 2 dies, it's bottle 2. If both die, it's bottle 3.

Reviewer: ```suggestion
If rat 1 dies, we know the poison is in the 1s place - either bottle 1 or 3 is poisoned.

If only rat 1 dies, it's bottle 1. If only rat 2 dies, it's bottle 2. If both die, it's bottle 3.
```

> To be clear: instead of testing bottles one by one, we will have the rats sip from _multiple_ bottles at once. Each rat's survival or death becomes a piece of information that we can _combine_ to locate the poisoned bottle.

Reviewer: ```suggestion
To be clear: Before we were linking rats to separate bottles, and only one rat told us information about any one bottle. And at most one rat would die per round. Instead, we will have the rats sip from _multiple_ bottles at once. Each rat's survival or death becomes a piece of information that we can _combine_ with information from other rates to locate the poisoned bottle.
```

Author: No they are still doing it in groups in the decimation round. I will make that clearer up there.
